File: src/features/ai-platform/infra/commands/proxy.ts
import { invoke } from '@tauri-apps/api/core';

export interface ProxyServerRecord {
  running: boolean;
  listenHost: string;
  listenPort: number;
  logRequests: boolean;
  connectTimeoutSeconds: number;
  requestTimeoutSeconds: number;
}

export interface ProxyAppSwitchesRecord {
  claude: boolean;
  codex: boolean;
  gemini: boolean;
  opencode: boolean;
  openclaw: boolean;
}

export interface FailoverPolicyRecord {
  enabled: boolean;
  consecutiveFailures: number;
  cooldownSeconds: number;
}

export interface ProxyControlPlaneRecord {
  server: ProxyServerRecord;
  takeover: ProxyAppSwitchesRecord;
  failover: ProxyAppSwitchesRecord;
  failoverPolicy: FailoverPolicyRecord;
  exposeProxyToggle: boolean;
  exposeFailoverToggle: boolean;
}

export interface ProxyAppStatus {
  app: string;
  running: boolean;
  takeoverEnabled: boolean;
  failoverEnabled: boolean;
  queueDepth: number;
  activeProviderId?: string;
  requiresProxy: boolean;
  status: string;
  detail: string;
}

export interface ProxyQueueItem {
  app: string;
  providerId: string;
  providerName: string;
  providerType: string;
  priority: number;
  isActive: boolean;
  requiresProxy: boolean;
}

export interface ProxySnapshot {
  config: ProxyControlPlaneRecord;
  appStatuses: ProxyAppStatus[];
  queue: ProxyQueueItem[];
  source: string;
}

export interface ProxyRuntimeStatus {
  running: boolean;
  host: string;
  port: number;
  message: string;
}

export function getAiPlatformProxySnapshot() {
  return invoke<ProxySnapshot>('ai_platform_get_proxy_snapshot');
}

export function saveAiPlatformProxyConfig(config: ProxyControlPlaneRecord) {
  return invoke<ProxySnapshot>('ai_platform_save_proxy_config', { config });
}

export function startAiPlatformProxy() {
  return invoke<ProxyRuntimeStatus>('ai_platform_start_proxy');
}

export function stopAiPlatformProxy() {
  return invoke<ProxyRuntimeStatus>('ai_platform_stop_proxy');
}

File: src-tauri/src/ai_platform/domain/proxy.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProxyServerRecord {
    pub running: bool,
    pub listen_host: String,
    pub listen_port: u16,
    pub log_requests: bool,
    pub connect_timeout_seconds: u32,
    pub request_timeout_seconds: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ProxyAppSwitchesRecord {
    pub claude: bool,
    pub codex: bool,
    pub gemini: bool,
    pub opencode: bool,
    pub openclaw: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FailoverPolicyRecord {
    pub enabled: bool,
    pub consecutive_failures: u32,
    pub cooldown_seconds: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProxyControlPlaneRecord {
    pub server: ProxyServerRecord,
    pub takeover: ProxyAppSwitchesRecord,
    pub failover: ProxyAppSwitchesRecord,
    pub failover_policy: FailoverPolicyRecord,
    pub expose_proxy_toggle: bool,
    pub expose_failover_toggle: bool,
}

impl Default for ProxyControlPlaneRecord {
    fn default() -> Self {
        Self {
            server: ProxyServerRecord {
                running: false,
                listen_host: "127.0.0.1".to_string(),
                listen_port: 8787,
                log_requests: false,
                connect_timeout_seconds: 15,
                request_timeout_seconds: 90,
            },
            takeover: ProxyAppSwitchesRecord::default(),
            failover: ProxyAppSwitchesRecord::default(),
            failover_policy: FailoverPolicyRecord {
                enabled: false,
                consecutive_failures: 3,
                cooldown_seconds: 60,
            },
            expose_proxy_toggle: true,
            expose_failover_toggle: false,
        }
    }
}